### PR TITLE
fix(27453): apply filters correctly in gatsby-source-drupal

### LINKED ISSUE #27453: Filters in gatsby-source-drupal aren't added to the first page of results
## Description

`gatsby-source-drupal` has the `filters` option which has an issue. The filters are only applied starting from the second page of the results. The first one doesn't get them. Example:
first call: `/jsonapi/node/page`
second call: `/jsonapi/node/page?page[offset]=50&page[limit]=50?filter=value`

### Steps to reproduce

- Add the gatsby-source-drupal plugin with the filters option (example: `filters: { 'node--page': 'filter=value'  }`).
- Run the build and check the urls that are hit.

### Expected result

- The filters should be added to the first page
- They shouldn't be added to the next pages (the link in `next.href` returned by jsonapi already has them)

### Actual result

See description.

### Environment

Run `gatsby info --clipboard` in your project directory and paste the output here.
```
  System:
    OS: Linux 4.15 Ubuntu 18.04.5 LTS (Bionic Beaver)
    CPU: (4) x64 Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
    Shell: 4.4.20 - /bin/bash
  Binaries:
    Node: 14.5.0 - ~/.nvm/versions/node/v14.5.0/bin/node
    Yarn: 1.19.0 - ~/.yarn/bin/yarn
    npm: 6.14.5 - ~/.nvm/versions/node/v14.5.0/bin/npm
  Languages:
    Python: 2.7.17 - /usr/bin/python
  Browsers:
    Chrome: 86.0.4240.75
    Firefox: 81.0
```

### PR DESCRIPTION
**Pull request by blazeyo**
## Description

Fixes issues with filter in gatsby-source-drupal described in #27453 

### Documentation

It's a bugfix, so no documentation changes are needed.

## Related Issues

Fixes #27453

### CODE REVIEW

File: packages/gatsby-source-drupal/src/gatsby-node.js [hunk 1]
```diff
@@ -220,19 +220,28 @@ exports.sourceNodes = async (
         if (disallowedLinkTypes.includes(type)) return
         if (!url) return
         if (!type) return
+
+        // Apply any filters configured in gatsby-config.js. Filters
+        // can be any valid JSON API filter query string.
+        // See https://www.drupal.org/docs/8/modules/jsonapi/filtering
+        if (typeof filters === `object`) {
```
Comment (Auspicus) [inline packages/gatsby-source-drupal/src/gatsby-node.js:227]: Is there any way to refactor this into a function that can be reused inside and before `getNext`?

Comment (blazeyo) [reply packages/gatsby-source-drupal/src/gatsby-node.js:227]: The recursive getNext call looks like this

```
if (d.data.links && d.data.links.next) {
  data = await getNext(d.data.links.next, data)
}
```

so it gets the subsequent pages for the same entity type listing. The `next` links already contain the filters which were applied to the first page, so there's no need to add them again.

Comment (Auspicus) [reply packages/gatsby-source-drupal/src/gatsby-node.js:227]: :+1: Perfect! Yeah I came back around to this and that's fine.

**Comment by KyleAMathews** [comment]
This issue should be fixed in https://github.com/gatsbyjs/gatsby/pull/31350

Let us know if not!

**Pull request closed without merging**